### PR TITLE
Cheap ordering of elements with Hilbert spatial sort [hilbert-sort-dev]

### PR DESCRIPTION
**Pull request by jakubcerveny**
Adds the method `Mesh::GetHilbertElementOrdering`, which performs a spatial sort of mesh elements, to be used for coarse mesh reordering. This is a cheap (but fast) alternative to `GetGeckoElementOrdering`, when preparing the mesh for NC AMR. Could also be useful to improve caching. 

The elements are simply mapped to the Hilbert curve on the unit square. The result is not perfect but it is way better than no ordering, and for many applications it may be sufficient. 

Example ordering:
hilbert-sort-0

When used for cheap partitioning (if we ever wanted to get rid of METIS as a mandatory dependency):

hilbert-sort-1
hilbert-sort-2

TODO:
* [x] Support for 3D meshes
* [x] Update CHANGELOG

| PR | Author | Editor | Reviewers  | Assignment | Approval | Merge |
| --- | --- | --- | ---  | --- | --- | --- |
| https://github.com/mfem/mfem/pull/1166 | @jakubcerveny | @v-dobrev | @rw-anderson + @bslazarov | 12/04/19 | 12/19/19 | ⌛due 12/28/19 |

### CODE REVIEW

**Comment by jakubcerveny** [comment]
Note: SFC-based partitioning tends to work better as the number of elements per process gets smaller:

hilbert-sort-3

**Comment by jakubcerveny** [comment]
Added support for 3D sorting.

hilbert-sort-6
hilbert-sort-4

**Comment by tzanio** [comment]
Oh, these are cool and may very well be sufficient for our purposes... @Dan2997925 can you give it a try in the problem where previously you had issues?

**Comment by rw-anderson** [comment]
Nice idea.  How is the scale of the Hilbert curve chosen?

**Comment by jakubcerveny** [comment]
The curve is mapped into the bounding box of the mesh. The question is whether this is OK or if the box should actually have equal sides. It seems to works OK on the pipe with a tight box...

**Comment by rw-anderson** [comment]
I was wondering more about how many recursion levels on the Hilbert curve there are within the bounding box.  Or maybe I am misunderstanding something.

**Comment by jakubcerveny** [comment]
There are two implementations actually, the first (2D only and disabled atm) converts a position into a distance along the curve (`PointToHilbertIndex`) and the elements are eventually sorted linearly by the index. This version uses 20 levels of the curve, or 20 bits per coordinate (60 in total for a potential 3D index). The second implementation does not construct the index explicitly and instead does something like a spatial QuickSort: starting at the top level, the points are partitioned into 8 octants by 3 midplanes. The function calls itself recursively for the octants (with some inversions and rotations of the coordinates), if they still contain some points. This version is only limited by the precision of the coordinates (53 bits?). I chose the second version even though the code is probably longer.

**Comment by jakubcerveny** [comment]
I think this is ready for review. The demo code in `ex1` and `ex1p` will be removed after review.

File: examples/ex1p.cpp [hunk 1]
```diff
@@ -122,19 +122,36 @@ int main(int argc, char *argv[])
...
+   if (mesh->NURBSext)
+   {
+      // reodering of NURBS meshes not supported by ReorderElements
+      mesh->SetCurvature(2);
+   }
+
+   Array<int> ordering;
+   mesh->GetHilbertElementOrdering(ordering);
+   mesh->ReorderElements(ordering);
+
```
Comment (tzanio) [inline examples/ex1p.cpp]: I like this interface... should be very easy for our users

File: mesh/mesh.cpp [hunk 2]
```diff
@@ -1383,6 +1383,177 @@ void Mesh::GetGeckoElementReordering(Array<int> &ordering,
...
+   HilbertCmp(int coord, bool dir, const Array<double> &points, double mid)
+      : coord(coord), dir(dir), points(points), mid(mid) {}
+
+   bool operator()(int i) const
+   {
+      return (points[3*i + coord] < mid) != dir;
+   }
+};
+
+static void HilbertSort2D(int coord1, bool dir1, bool dir2,
```
Comment (rw-anderson) [inline mesh/mesh.cpp]: Nice algorithm here.  Looks like dir1 and dir2 relate to rotating and flipping the quadrants upon recursion, but their meaning isn't that obvious at the outset.  I don't know if there's a comment that would make it obvious or not.

**Comment by v-dobrev** [comment]
We should probably add a menu option in the `mesh-explorer` miniapp to apply this reordering. The code there can serve as illustration for users. The Gecko-based reordering can be added there too.

**Comment by jakubcerveny** [comment]
@bslazarov Thanks for testing. I will add an ordering option in `mesh-explorer`.

The algorithm should scale very well. This is basically QuickSort, so something like O(n*log(n)). You can use this on millions of elements -- whatever fits in memory.

**Comment by tzanio** [comment]
@bslazarov and @jakubcerveny -- should we wait for the ordering option in `mesh-explorer` or is this approved and ready for `next` as it is?

File: mesh/mesh.cpp (hunk 2 above)
Comment (jakubcerveny) [reply mesh/mesh.cpp]: Added a couple comments related to this.

**Comment by jakubcerveny** [comment]
Added a new submenu to reorder the mesh in `mesh-explorer`. Gecko will go there in the future too. Also added the option to partition the mesh along the 1D sequence of all elements.

This was done in `mesh-expolorer` using the above:
explorer

**Comment by tzanio** [comment]
@jakubcerveny, can you mention the new `mesh-explorer` option in `CHANGELOG`?

@bslazarov, can you take another look and approve?

**Comment by tzanio** [comment]
Merged in `next` for testing ...

**Comment by tzanio** [comment]
Unit testing fails as shown below. Can you fix please.

See the [full log](https://mybitbucket.llnl.gov/projects/MFEM/repos/autotest/commits/19cd0772c751644e263ffebe335fc77c3276c084#tux426/2019-12-20-llnl-gh-next/unit-test.out)) in mfem/autotest.

```
Building MFEM with Gecko ...
Building the unit tests ...
rm -f *.o */*.o */*~ *~
g++ -g -c unit_test_main.cpp -I.  -g -Wall -std=c++11 -I/home/kolev1/autotest-mfem/tux426/mfem -I/home/kolev1/autotest-mfem/tux426/mfem/../gecko/inc -o unit_test_main.o
g++ -g -c fem/test_1d_bilininteg.cpp -I.  -g -Wall -std=c++11 -I/home/kolev1/autotest-mfem/tux426/mfem -I/home/kolev1/autotest-mfem/tux426/mfem/../gecko/inc -o fem/test_1d_bilininteg.o
g++ -g -c fem/test_2d_bilininteg.cpp -I.  -g -Wall -std=c++11 -I/home/kolev1/autotest-mfem/tux426/mfem -I/home/kolev1/autotest-mfem/tux426/mfem/../gecko/inc -o fem/test_2d_bilininteg.o
g++ -g -c fem/test_3d_bilininteg.cpp -I.  -g -Wall -std=c++11 -I/home/kolev1/autotest-mfem/tux426/mfem -I/home/kolev1/autotest-mfem/tux426/mfem/../gecko/inc -o fem/test_3d_bilininteg.o
g++ -g -c fem/test_calcshape.cpp -I.  -g -Wall -std=c++11 -I/home/kolev1/autotest-mfem/tux426/mfem -I/home/kolev1/autotest-mfem/tux426/mfem/../gecko/inc -o fem/test_calcshape.o
g++ -g -c fem/test_datacollection.cpp -I.  -g -Wall -std=c++11 -I/home/kolev1/autotest-mfem/tux426/mfem -I/home/kolev1/autotest-mfem/tux426/mfem/../gecko/inc -o fem/test_datacollection.o
g++ -g -c fem/test_fe.cpp -I.  -g -Wall -std=c++11 -I/home/kolev1/autotest-mfem/tux426/mfem -I/home/kolev1/autotest-mfem/tux426/mfem/../gecko/inc -o fem/test_fe.o
g++ -g -c fem/test_intrules.cpp -I.  -g -Wall -std=c++11 -I/home/kolev1/autotest-mfem/tux426/mfem -I/home/kolev1/autotest-mfem/tux426/mfem/../gecko/inc -o fem/test_intrules.o
g++ -g -c fem/test_intruletypes.cpp -I.  -g -Wall -std=c++11 -I/home/kolev1/autotest-mfem/tux426/mfem -I/home/kolev1/autotest-mfem/tux426/mfem/../gecko/inc -o fem/test_intruletypes.o
g++ -g -c fem/test_inversetransform.cpp -I.  -g -Wall -std=c++11 -I/home/kolev1/autotest-mfem/tux426/mfem -I/home/kolev1/autotest-mfem/tux426/mfem/../gecko/inc -o fem/test_inversetransform.o
g++ -g -c fem/test_lin_interp.cpp -I.  -g -Wall -std=c++11 -I/home/kolev1/autotest-mfem/tux426/mfem -I/home/kolev1/autotest-mfem/tux426/mfem/../gecko/inc -o fem/test_lin_interp.o
g++ -g -c fem/test_linear_fes.cpp -I.  -g -Wall -std=c++11 -I/home/kolev1/autotest-mfem/tux426/mfem -I/home/kolev1/autotest-mfem/tux426/mfem/../gecko/inc -o fem/test_linear_fes.o
g++ -g -c fem/test_quadraturefunc.cpp -I.  -g -Wall -std=c++11 -I/home/kolev1/autotest-mfem/tux426/mfem -I/home/kolev1/autotest-mfem/tux426/mfem/../gecko/inc -o fem/test_quadraturefunc.o
g++ -g -c linalg/test_blockMatrix.cpp -I.  -g -Wall -std=c++11 -I/home/kolev1/autotest-mfem/tux426/mfem -I/home/kolev1/autotest-mfem/tux426/mfem/../gecko/inc -o linalg/test_blockMatrix.o
g++ -g -c linalg/test_densematrix.cpp -I.  -g -Wall -std=c++11 -I/home/kolev1/autotest-mfem/tux426/mfem -I/home/kolev1/autotest-mfem/tux426/mfem/../gecko/inc -o linalg/test_densematrix.o
g++ -g -c mesh/test_mesh.cpp -I.  -g -Wall -std=c++11 -I/home/kolev1/autotest-mfem/tux426/mfem -I/home/kolev1/autotest-mfem/tux426/mfem/../gecko/inc -o mesh/test_mesh.o
mesh/test_mesh.cpp: In function ‘void ____C_A_T_C_H____T_E_S_T____0()’:
mesh/test_mesh.cpp:19:15: error: ‘class mfem::Mesh’ has no member named ‘GetGeckoElementReordering’
          mesh.GetGeckoElementReordering(perm);
               ^
mesh/test_mesh.cpp:41:15: error: ‘class mfem::Mesh’ has no member named ‘GetGeckoElementReordering’
          mesh.GetGeckoElementReordering(perm);
               ^
mesh/test_mesh.cpp:65:22: error: ‘class mfem::Mesh’ has no member named ‘GetGeckoElementReordering’
       mesh_reordered.GetGeckoElementReordering(perm);
                      ^
make: *** [mesh/test_mesh.o] Error 1
make: *** Waiting for unfinished jobs....
Building the unit test code failed. Stop
```

File: examples/ex1.cpp [hunk 3]
```diff
@@ -112,6 +112,7 @@ int main(int argc, char *argv[])
       }
    }
 
+#if 0
```
Comment (tzanio) [inline examples/ex1.cpp]: Can you please revert these and re-merge in `next`?

Comment (jakubcerveny) [reply examples/ex1.cpp]: Done, but I can't push to `next`:
```
remote: error: You're not authorized to push to this branch.
```

**Comment by jakubcerveny** [comment]
Regarding the unit tests, this is because I renamed `GetGeckoElementReordering` -> `GetGeckoElementOrdering`. I did update the unit tests in this branch, but somehow the old tests are being run.

File: examples/ex1.cpp (hunk 3 above)
Comment (tzanio) [reply examples/ex1.cpp]: Ok, I pushed. Can you make sure you are in the LLNL org + you have 2FA enabled? That's required to be able to push to `next` now because of GitLab.

**Comment by tzanio** [comment]
> Regarding the unit tests, this is because I renamed `GetGeckoElementReordering` -> `GetGeckoElementOrdering`. I did update the unit tests in this branch, but somehow the old tests are being run.

That's weird. I updated the internal repo... hopefully should be OK now.

**Comment by tzanio** [comment]
Re-merged in `next` for testing ...

File: CHANGELOG [hunk 4]
```diff
@@ -56,6 +56,23 @@ Miscellaneous
...
+
+- New options to reorder and partition the mesh in the mesh-explorer miniapp.
+
+- Improved element numbering after uniform mesh refinement.
+
+New and updated examples and miniapps
+-------------------------------------
+- The (p)mesh-optimizer miniapp has been updated to demonstrate mesh optimization
+  for an AMR mesh.
+
```
Comment (v-dobrev) [inline CHANGELOG]: Some of the above text comes from different PRs and was merged here incorrectly -- e.g. there are bullets that are repeated twice.

File: mesh/mesh.cpp [hunk 5]
```diff
@@ -1345,9 +1345,9 @@ void Mesh::FinalizeQuadMesh(int generate_edges, int refine,
 
 
 #ifdef MFEM_USE_GECKO
-void Mesh::GetGeckoElementReordering(Array<int> &ordering,
-                                     int iterations, int window,
-                                     int period, int seed)
+void Mesh::GetGeckoElementOrdering(Array<int> &ordering,
```
Comment (v-dobrev) [inline mesh/mesh.cpp]: Do not rename things that are already in the library -- I think you should revert this change.

File: CHANGELOG (hunk 4 above)
Comment (tzanio) [reply CHANGELOG]: This is fixed now

File: mesh/mesh.cpp (hunk 5 above)
Comment (tzanio) [reply mesh/mesh.cpp]: I generally agree, but I think the change is acceptable and desirable in this particular case

**Comment by tzanio** [comment]
Re-merged in `next` for testing ...

File: mesh/mesh.cpp (hunk 5 above)
Comment (jakubcerveny) [reply mesh/mesh.cpp]: Sorry for the late reply. I agree we shouldn't just change things in the API, but this case I felt was special. Since the library was never published, this code was hardly ever used by anyone outside the core team. That's why I thought the change was acceptable in this case.

File: miniapps/meshing/mesh-explorer.cpp [hunk 6]
```diff
@@ -221,8 +220,10 @@ int main (int argc, char *argv[])
            "x) Print sub-element stats\n"
            "f) Find physical point in reference space\n"
            "p) Generate a partitioning\n"
+           "o) Reorder elements\n"
```
Comment (tzanio) [inline miniapps/meshing/mesh-explorer.cpp]: Besides `e) View elements` is there a better way to visualize the ordering?

Comment (jakubcerveny) [reply miniapps/meshing/mesh-explorer.cpp]: I don't know of a direct way. I was just using the new "1D split" partitioning to visualize the partitions. The "curve" plot in the PR description is from Octave.
